File: src/e2e/rtmp_auth_test.go
//go:build e2e

package e2e

import (
	"fmt"
	"os"
	"path/filepath"
	"testing"
	"time"

	"Theatrum/constants"
)

func TestRTMPAuth_ValidToken(t *testing.T) {
	liveStreamKey := "auth-test-secret"
	username := "validuser"
	token := computeXORToken(liveStreamKey, username)

	ts := setupTestServer(t, TestConfig{
		Channels: fmt.Sprintf(`  "/user/{username}":
    stream:
      type: live
      path: "live/{username}"
      live_stream_key: "%s"
      auth_token_template: "{username}"
      distribution:
        hls:
          segment_duration: 2
          window_size: 3
`, liveStreamKey),
	})

	// Start RTMP stream with valid token
	rtmpURL := fmt.Sprintf("rtmp://127.0.0.1:%d/user/%s", ts.RTMPPort, username)
	streamToRTMP(t, rtmpURL, token, 10)

	// Verify that HLS files are created (auth succeeded)
	playlistPath := filepath.Join(ts.DataDir, "live", username, constants.DefaultQuality, constants.SubPlaylist)
	waitForFile(t, playlistPath, 15*time.Second)

	t.Log("Valid auth token: stream started successfully")
}

func TestRTMPAuth_InvalidToken(t *testing.T) {
	liveStreamKey := "auth-test-secret"
	username := "baduser"

	ts := setupTestServer(t, TestConfig{
		Channels: fmt.Sprintf(`  "/user/{username}":
    stream:
      type: live
      path: "live/{username}"
      live_stream_key: "%s"
      auth_token_template: "{username}"
      distribution:
        hls:
          segment_duration: 2
          window_size: 3
`, liveStreamKey),
	})

	// Start RTMP stream with invalid token
	rtmpURL := fmt.Sprintf("rtmp://127.0.0.1:%d/user/%s", ts.RTMPPort, username)
	cmd := streamToRTMP(t, rtmpURL, "invalid-token", 5)

	// FFmpeg should exit with error since authentication fails
	done := make(chan error, 1)
	go func() {
		done <- cmd.Wait()
	}()

	select {
	case err := <-done:
		// FFmpeg should fail (auth rejected)
		if err == nil {
			t.Log("FFmpeg exited cleanly - auth may have silently failed")
		} else {
			t.Logf("FFmpeg exited with error (expected): %v", err)
		}
	case <-time.After(10 * time.Second):
		t.Log("FFmpeg did not exit quickly, auth may have failed at different stage")
		cmd.Process.Kill()
	}

	// Verify no HLS files were created
	time.Sleep(2 * time.Second)
	playlistPath := filepath.Join(ts.DataDir, "live", username, constants.DefaultQuality, constants.SubPlaylist)
	if _, err := os.Stat(playlistPath); err == nil {
		t.Error("Expected no HLS files for invalid token, but playlist was created")
	}
}

func TestRTMPAuth_UnknownChannel(t *testing.T) {
	liveStreamKey := "auth-test-secret"

	ts := setupTestServer(t, TestConfig{
		Channels: fmt.Sprintf(`  "/user/{username}":
    stream:
      type: live
      path: "live/{username}"
      live_stream_key: "%s"
      auth_token_template: "{username}"
      distribution:
        hls:
          segment_duration: 2
          window_size: 3
`, liveStreamKey),
	})

	// Try to connect to a channel that doesn't exist
	rtmpURL := fmt.Sprintf("rtmp://127.0.0.1:%d/nonexistent/channel", ts.RTMPPort)
	cmd := streamToRTMP(t, rtmpURL, "anytoken", 5)

	// Wait for FFmpeg to exit (should fail)
	done := make(chan error, 1)
	go func() {
		done <- cmd.Wait()
	}()

	select {
	case <-done:
		// Expected: FFmpeg exits because server rejects the unknown channel
	case <-time.After(10 * time.Second):
		cmd.Process.Kill()
	}

	// Verify no files in data dir for this path
	time.Sleep(1 * time.Second)
	unknownDir := filepath.Join(ts.DataDir, "nonexistent")
	if _, err := os.Stat(unknownDir); err == nil {
		t.Error("Expected no files for unknown channel")
	}
}

File: src/test/java/com/fjourdren/theatrum/e2e/RtmpAuthTest.java
package com.fjourdren.theatrum.e2e;

import com.fjourdren.theatrum.domain.constant.VideoConstants;
import org.junit.jupiter.api.AfterAll;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.io.TempDir;

import java.nio.file.Files;
import java.nio.file.Path;
import java.time.Duration;
import java.util.concurrent.TimeUnit;

import static org.assertj.core.api.Assertions.assertThat;

/** Port of Go's {@code rtmp_auth_test.go}: XOR token authentication end to end. */
class RtmpAuthTest {

    private static final String KEY = "auth-test-secret";

    private static final String CHANNELS = """
              "/user/{username}":
                stream:
                  type: live
                  path: "live/{username}"
                  live_stream_key: "%s"
                  auth_token_template: "{username}"
                  distribution:
                    hls:
                      segment_duration: 2
                      window_size: 3
            """.formatted(KEY);

    @TempDir
    static Path tempDir;

    static TestServerSupport server;

    @BeforeAll
    static void startServer() {
        TestServerSupport.requireFfmpeg();
        server = TestServerSupport.start(tempDir, CHANNELS);
    }

    @AfterAll
    static void stopServer() {
        if (server != null) {
            server.close();
        }
    }

    @Test
    void tokenIsTheHexOfTheAuthInputXoredWithTheKey() {
        // Worked example, verified against the running server.
        assertThat(TestServerSupport.xorToken("secret", "alice")).isEqualTo("12090a1100");
    }

    @Test
    void acceptsAValidToken() {
        String username = "validuser";
        Process publisher = server.publish("user/" + username, TestServerSupport.xorToken(KEY, username), 10);

        Path playlist = server.dataDir.resolve("live").resolve(username)
                .resolve(VideoConstants.DEFAULT_QUALITY).resolve(VideoConstants.SUB_PLAYLIST);
        TestServerSupport.waitForFile(playlist, Duration.ofSeconds(20));

        TestServerSupport.kill(publisher);
    }

    @Test
    void rejectsAnInvalidToken() throws Exception {
        String username = "baduser";
        Process publisher = server.publish("user/" + username, "invalid-token", 5);

        // FFmpeg is dropped at publish time, so it exits well before its -t duration.
        assertThat(publisher.waitFor(15, TimeUnit.SECONDS)).as("FFmpeg should be rejected").isTrue();

        assertThat(server.dataDir.resolve("live").resolve(username)
                .resolve(VideoConstants.DEFAULT_QUALITY).resolve(VideoConstants.SUB_PLAYLIST))
                .doesNotExist();
    }

    @Test
    void rejectsAnUnknownChannel() throws Exception {
        Process publisher = server.publish("nonexistent/channel", "anytoken", 5);

        assertThat(publisher.waitFor(15, TimeUnit.SECONDS)).as("FFmpeg should be rejected").isTrue();
        assertThat(Files.exists(server.dataDir.resolve("nonexistent"))).isFalse();
    }
}
